### ISSUE TITLE
Support scram-sha-256 authentication

### ISSUE THREAD

**Issue by reonZ** [denodrivers/postgres#182]
As soon as I attempt a `Client.connect` I am having this error:

```
error: Uncaught Error: Unknown auth message code 10
        throw new Error('Unknown auth message code ${code}');
              ^
    at Connection.handleAuth (https://deno.land/x/postgres@v0.4.5/connection.ts:197:15)
    at Connection.startup (https://deno.land/x/postgres@v0.4.5/connection.ts:155:16)
    at async Client.connect (https://deno.land/x/postgres@v0.4.5/client.ts:14:5)
```

my client instantiation:
```
const client = new Client({
    user: 'postgres',
    password: 'postgres',
    database: 'test',
    hostname: 'localhost',
    port: 5432,
})
```

The permissions at launch `deno run --no-check --unstable --allow-net --allow-env --allow-read app.ts --local`

The service is obviously running and I can access the database through other means with the same credentials.

Edit: It would seems like i am able to connect to a remote database, but not the local one.

**Comment by ebebbington** [comment]
You sure your local db uses those same credentials as authentication?

**Comment by hidjou** [comment]
I'm getting this error too, but it seems to only happen when I use Postgres 13, if I switch back to 12 it works just fine

**Comment by EliSauder** [comment]
The error is related to supported encryption types. Newer versions of postgresql use scram-sha-256 for authentication. However, it appears that this driver only supports md5 encryption. To use new versions you will have to manually enter in a md5 hash for your postgres user's password.
How to generate md5 authentication for postgres: https://stackoverflow.com/a/58979721
Once you have the hash you can do:
```sql
ALTER USER username WITH ENCRYPTED PASSWORD 'md5yourmd5hashhere';
```

This driver should be updated to use the newer authentication as it is more secure.

**Comment by sanjay4one** [comment]
@EliSauder ,
I am also facing the same issue. I tried with your solution but it's not working for me.

First I created the the md5 hash by using the below command:

`echo -n "md5"; md5 -qs "admin123sanjay"` (user: sanjay, password: admin123)
`md578d0addbb4c5f309367b66a6f5f0cc3d`

And below is the postgres connection request:

`export const client = new Client({
    user: "sanjay",
    database: "simple_shop",
    hostname: "localhost",
    password: "md578d0addbb4c5f309367b66a6f5f0cc3d",
    port: 5432,
})`

But still getting below error:

`error: Uncaught (in promise) Error: Unknown auth message code 10
        throw new Error(`Unknown auth message code ${code}`);
              ^
    at Connection.handleAuth (connection.ts:197:15)
    at Connection.startup (connection.ts:155:16)
    at async Client.connect (client.ts:14:5)
    at async index.ts:11:1`

Please let me know how to fix this. Thanks in advance.

**Comment by EliSauder** [comment]
@sanjay4one 
In your client creation, use the actual password "admin123". You just need to make sure that postgress is storing the password using the md5 hash. Behind the scenes postgress is combining the user name and password and regenerating the hash to see if it matches the one you saved.

Steps:
1. Create a user with an md5 encrypted password (in your case, md578d0addbb4c5f309367b66a6f5f0cc3d).
2. Log in using the user name and password (in your case, user: sanjay, pass: admin123)
   * Behind the scenes postgress will hash your username and password together and validate that against the one you created the user with.

**Comment by sanjay4one** [comment]
@EliSauder 
Thank you for your reply. Actual password is also not working. Please verify the below connection request:

`export const client = new Client({
    user: "sanjay",
    database: "simple_shop",
    hostname: "localhost",
    password: "admin123",
    port: 5432,
})`

And the stored user information is:

`CREATE ROLE sanjay WITH
  LOGIN
  SUPERUSER
  INHERIT
  CREATEDB
  CREATEROLE
  REPLICATION
  ENCRYPTED PASSWORD 'md578d0addbb4c5f309367b66a6f5f0cc3d';`

Thank you!!

**Comment by EliSauder** [comment]
@sanjay4one 
Everything there looks good. I actually also worked through it on my end to see if it works and I was able to get it to work using the following:

(Note my server uses port 5433 instead of 5432)

Using your query: 
```SQL
CREATE ROLE sanjay WITH LOGIN SUPERUSER INHERIT CREATEDB CREATEROLE REPLICATION ENCRYPTED PASSWORD 'md578d0addbb4c5f309367b66a6f5f0cc3d';
```
And the following connection:
```ts
const client = new Client({
    user: "sanjay",
    password: "admin123",
    database: "testDB",
    hostname: "localhost",
    port: 5433
  });
```

I think the issue you are having now is unrelated to the login and more on which server/db you are trying to connect to

**Comment by Soremwar** [comment]
I've encountered this error in the past, and all the times it was related to the particular postgres server config I was using rather than the library it was using as a client

Closing this for now, unless you have proof that this is indeed a problem of the library

**Comment by EliSauder** [comment]
@Soremwar This isn't so much a bug in the library as it is a feature that does not exist (not supporting the newer scram-sha-256 authentication method. It can be mitigated by manually specifying your authentication type to md5 on the server in the pg_hba.conf file. However postgress comments that scram-sha-256 is more secure as seen in the postgress documentation [20.5. Password Authentication](https://www.postgresql.org/docs/13/auth-password.html). 

The code in question is seen below from [connection.handleAuth](https://github.com/denodrivers/postgres/blob/5472688a1caf9aa8e8861c4303a8d6148e2b6645/connection.ts#L183)

```ts
  async handleAuth(msg: Message) {
    const code = msg.reader.readInt32();
    switch (code) {
      case 0:
        // pass
        break;
      case 3:
        // cleartext password
        await this._authCleartext();
        await this._readAuthResponse();
        break;
      case 5: {
        // md5 password
        const salt = msg.reader.readBytes(4);
        await this._authMd5(salt);
        await this._readAuthResponse();
        break;
      }
      default:
        throw new Error(`Unknown auth message code ${code}`);
    }
  }
```

**Comment by Soremwar** [comment]
If this is a missing feature from the library then I'm interested in supporting it

I'll look into this and probably implement it into a minor release sometime in the future

**Comment by antoniogamiz** [comment]
Any updates on this? :D

**Comment by snsinfu** [comment]
I know the issue has already been assigned, but for my own needs I implemented scram-sha-256 auth [in my fork](https://github.com/snsinfu/deno-postgres/tree/feat-scram). Import:

```ts
import { Client } from "https://raw.githubusercontent.com/snsinfu/deno-postgres/feat-scram/mod.ts";
```

**Comment by Soremwar** [comment]
This is huge @snsinfu . I didn't knew much about the SASL protocol myself, so I had been posponing this for a while now. Care to open a PR? Can't believe you even added tests, this is crazy

**Comment by snsinfu** [comment]
@Soremwar Sure! I opened a PR: #267.

**Comment by Soremwar** [comment]
Thanks to the effort of @snsinfu this is now supported, you can take it for a ride before the 0.9.0 release by importing from `https://raw.githubusercontent.com/denodrivers/postgres/master/mod.ts`

Closed in #267
